onchain-development local-setup updated (#343)
solana-foundation/developer-content: content/courses/onchain-development/local-setup.md

## Lesson

This lesson is a guide to installing the tools required for developing onchain
programs. Let's install Solana CLI tools, the Rust SDK, and Anchor, and create a
test program to ensure that our setup works.

## Lab

### Extra steps for Windows users

> > MacOS and Linux users can skip this section. Linux and MacOS are the

Reviewer: It's `macOS` see https://www.apple.com/macos/sonoma/

> recommended operating systems for Solana program development. However, if
> you're on Windows, you can still follow along with these extra steps.

Firstly, make sure you have Windows Terminal installed, otherwise you can

> install it from the
> [Microsoft store](https://apps.microsoft.com/detail/9N0DX20HK701).

Reviewer: ```suggestion
[install Windows Terminal from the
[Microsoft store](https://apps.microsoft.com/detail/9N0DX20HK701).
```

Just making the link text more obvious

Then,
[install Windows Subsystem for Linux (WSL)](https://learn.microsoft.com/en-us/windows/wsl/install).
WSL provides a Linux environment that launches instantly whenever you need it
and doesn't slow your computer down.

Start Windows Terminal, launch an 'Ubuntu' session inside the terminal, and
proceed with the rest of these steps.

### Download Rust

First, download Rust by
[following the instructions](https://www.rust-lang.org/tools/install):

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

### Download the Solana CLI tools

Next,
[download the Solana CLI tools](https://docs.solana.com/cli/install-solana-cli-tools).

> > the command below will install the stable version of the Solana CLI.
> 
> ```bash
> sh -c "$(curl -sSfL https://release.anza.xyz/stable/install)"
> ```

Reviewer 2: Installing Solana via channel names is no longer supported (see [here](https://github.com/solana-labs/solana/releases)).

```suggestion
Solana installation no longer supports symbolic channel names (`edge`, `beta`, `stable`), hence we'll have to specify the version.

> the command below will install version `1.18.22` of the Solana CLI.

```bash
sh -c "$(curl -sSfL https://release.anza.xyz/v1.18.22/install)"
```
```

Afterwards, `solana -V` should show `solana-cli 1.18.x` (any number for `x` is
fine) which is the version of the Solana CLI installed in your system indicating

> that it was installed successfully.

Reviewer: Suggest tweaking this to:
---
Afterwards, run `solana -V` to see the version of the Solana CLI installed on your system. 

(show the output of solana -V)

This should show `solana-cli 1.18.x` - any number for `x` is fine. If so, the Solana CLI was installed successfully!

### Download Anchor

Finally, [download Anchor](https://www.anchor-lang.com/docs/installation):

```bash
cargo install --git https://github.com/coral-xyz/anchor avm --locked --force
```

you may need to install additional dependencies in Linux (or WSL):

```bash
sudo apt-get update && \
sudo apt-get upgrade && \
sudo apt-get install -y pkg-config build-essential libudev-dev libssl-dev
```

proceed...

```bash
avm install latest
avm use latest
```

Afterwards, `anchor -V` should show `anchor-cli 0.30.x` (any number for `x` is
fine) which is the version of the Solana CLI installed in your system indicating
that it was installed successfully.

> ### Initialize Anchor project

Reviewer: ```suggestion
### Create an Anchor project
```

'Initialize' is vague.

## Lesson

This lesson is a guide to installing the tools required for developing onchain
programs. Let's install Solana CLI tools, the Rust SDK, and Anchor, and create a
test program to ensure that our setup works.

## Lab

### Extra steps for Windows users

> macOS and Linux users can skip this section. Linux and macOS are the
> recommended operating systems for Solana program development. However, if

> > you're on Windows, you can still follow along with these extra steps.

Reviewer: I'm not completely comfortable with:

> Linux and macOS are the
> recommended operating systems for Solana program development

@nickfrosty can you please give a second opinion here?

Author: I thought it may be appropriate to add this, as on Windows you still need to get WSL for local setup, which is in a way setting up a Linux-like environment for it. Whereas, you can just get started much more quickly and easily.
Personally, I use Linux and Windows, and the experience is better on Linux hands down. 
And I've met a handful of devs in the ecosystem who prefers mac or linux

Author: Though leaving it to users could be better choice. 
Should I fix this part?

Reviewer: We don't want Windows users to feel that they need to purchase an expensive Apple laptop or learn Linux just to program Solana. Just remove it. They already know these steps only apply to them.

Author: sure

Firstly, make sure you have Windows Terminal installed, otherwise you can
install it from the
[Microsoft store](https://apps.microsoft.com/detail/9N0DX20HK701).

Then,
[install Windows Subsystem for Linux (WSL)](https://learn.microsoft.com/en-us/windows/wsl/install).
WSL provides a Linux environment that launches instantly whenever you need it
and doesn't slow your computer down.

Start Windows Terminal, launch an 'Ubuntu' session inside the terminal, and
proceed with the rest of these steps.

### Download Rust

First, download Rust by
[following the instructions](https://www.rust-lang.org/tools/install):

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

### Download the Solana CLI tools

Next,
[download the Solana CLI tools](https://docs.solana.com/cli/install-solana-cli-tools).

Solana installation no longer supports symbolic channel names (`edge`, `beta`,

> `stable`), hence we'll have to specify the version.

Reviewer: This doesn't seem true:

```
% sh -c "$(curl -sSfL https://release.anza.xyz/stable/install)"
downloading stable installer
Install is up to date. e5d267d is the latest commit for stable
```

Author: @jereldlimjy  suggested this change. 
https://github.com/solana-labs/solana/releases

Reviewer: @SAMAD101 Yep, it fails if you use the old link, anza.xyz is good though.

Author: I'll revert the change then

> the command below will install version `1.18.22` of the Solana CLI.

```bash
sh -c "$(curl -sSfL https://release.anza.xyz/v1.18.22/install)"
```

Afterwards, `solana -V` should show `solana-cli 1.18.x` (any number for `x` is
fine) which is the version of the Solana CLI installed in your system indicating
that it was installed successfully.

### Download Anchor

Finally, [download Anchor](https://www.anchor-lang.com/docs/installation):

```bash
cargo install --git https://github.com/coral-xyz/anchor avm --locked --force
```

> 

Reviewer: Do we know what error message people would get if they didn't have these dependencies? Can we add it?

Author: I'll spin up a VM to see what errors one may get.

Author: I checked in VM. The error was not specific to anchor or solana cli. It was more of linux and rust related. The compilation would fail without those dependencies, as those are important dev dependencies on linux.

Reviewer: > The compilation would fail without those dependencies

That's what I suspected - so `anchor build` would throw an error? Write what the error is.

Reviewer: Going to resolve this, as someone else already won the bounty for this lesson and these changes are still excellent editions.

```

**The `anchor test` command should complete with no errors or warnings**.

**However you may encounter issues, and we'll fix them below:**

- `package `solana-program
  v1.18.12` cannot be built because it requires rustc 1.75.0 or newer` error

This error is due to incompatible versions of `solana-program` and `solana-cli`.
Run `cargo add solana-program@"=1.18.x"`, where `x` matches your version of
`solana-cli`. Then re-run `anchor test`.

> - `Error: Unable to read keypair file`

Reviewer: Please use headings (I think `####` not 5) that way we can deep link individual sections.

Add a keypair to `.config/solana/id.json`. You can either copy a keypair from an
`.env` file (just the array of numbers) into a file or use the command
`solana-keygen new --no-bip39-passphrase` to create a new keypair file. Then
re-run `anchor test`.

- `unused variable: 'ctx'` warning

This simply means the `initialize` instruction handler isn't doing anything yet.
You can open `programs/favorites/src/lib.rs` and change `ctx` to `_ctx` or just
go onto the next step.

> - `No license field in package.json` warning

Reviewer: A couple of other ones to add:

```
#### error: no such command: `build-sbf`

export PATH=~/.local/share/solana/install/active_release/bin:$PATH  Add this to your ~/.zshrc or ~/.bashrc file to make the change permanent.

### Unable to get latest blockhash. Test validator does not look started. 

Install Homebrew and use it to install GNU tar:  brew install gnu-tar
export PATH=/opt/homebrew/opt/gnu-tar/libexec/gnubin:$PATH Add this to your ~/.zshrc or ~/.bashrc file to make the change permanent.

```

(please format nicely per above)

Author: got it

Author: should I add some explanation - like their cause - for the errors listed in the doc?

```

**The `anchor test` command should complete with no errors or warnings**.

**However you may encounter issues, and we'll fix them below:**

#### `package `solana-program

v1.18.12` cannot be built because it requires rustc 1.75.0 or newer` error

This error is due to incompatible versions of `solana-program` and `solana-cli`.
Run `cargo add solana-program@"=1.18.x"`, where `x` matches your version of
`solana-cli`. Then re-run `anchor test`.

#### `Error: Unable to read keypair file`

Add a keypair to `.config/solana/id.json`. You can either copy a keypair from an
`.env` file (just the array of numbers) into a file or use the command
`solana-keygen new --no-bip39-passphrase` to create a new keypair file. Then
re-run `anchor test`.

#### error: no such command: `build-sbf`

> This error typically occurs because the Solana Build SBF (Solana BPF) feature is
> not available in your current environment. Solana BPF is used for compiling
> programs to run on the Solana blockchain. The error might indicate that either
> the Solana installation is corrupted or the relevant binaries are not in your
> system's PATH variable.
> 
> One possible solution is to add the Solana binaries to your system's PATH
> variable. Add this to your `~/.zshrc` or `~/.bashrc` file to make the change
> permanent.

Reviewer: ```suggestion
If you see this message, this error typically occurs because the relevant binaries are not in your shell's PATH variable.

Run this command to add this folder to your shell, and also add this to your `~/.zshrc` or `~/.bashrc` file to make the change permanent.
```

Reviewer: A few comments to help with future PRs (because your work is good and we want to do more in future):

 - 'system' is vague. Specifically, the thing we're fixing here is the shell. So let's say that!
 - 'Install is corrupted' is vague and kind of scary. How would an install be corrupted? The error is just that the shell needs a folder added to PATH. So let's be specific!

```bash
export PATH=~"/.local/share/solana/install/active_release/bin:$PATH"
```

> #### Unable to get latest blockhash. Test validator does not look started.
> 
> - Install [Homebrew](https://brew.sh/) and use it to install GNU tar:

Reviewer: ```suggestion
#### Unable to get latest blockhash. Test validator does not look started.

There's multiple versions of the 'tar' (tape archiver) command Solana used for archiving. macOS comes with BSD tar, but Solana CLI wants the GNU version installed.

- Install [Homebrew](https://brew.sh/) and use it to install GNU tar:
```

```

**The `anchor test` command should complete with no errors or warnings**.

**However you may encounter issues, and we'll fix them below:**

#### `package `solana-program

v1.18.12` cannot be built because it requires rustc 1.75.0 or newer` error

This error is due to incompatible versions of `solana-program` and `solana-cli`.
Run `cargo add solana-program@"=1.18.x"`, where `x` matches your version of
`solana-cli`. Then re-run `anchor test`.

#### `Error: Unable to read keypair file`

Add a keypair to `.config/solana/id.json`. You can either copy a keypair from an
`.env` file (just the array of numbers) into a file or use the command
`solana-keygen new --no-bip39-passphrase` to create a new keypair file. Then
re-run `anchor test`.

#### error: no such command: `build-sbf`

> 

Reviewer: It's worth adding an explanation for why the error is occurring. Eg "after installing Solana CLI, you'll need to ensure it's executables are in your PATH"...

Add this to your `~/.zshrc` or `~/.bashrc` file to make the change permanent.

```bash
export PATH=~"/.local/share/solana/install/active_release/bin:$PATH"
```

#### Unable to get latest blockhash. Test validator does not look started.

> - Install Homebrew and use it to install GNU tar:

Reviewer: Add a link to Homebrew project.
